File: src/Homespun.Shared/Models/Sessions/TodoParser.cs
using System.Text.Json;

namespace Homespun.Shared.Models.Sessions;

/// <summary>
/// Interface for parsing todo items from Claude session messages.
/// </summary>
public interface ITodoParser
{
    /// <summary>
    /// Extracts the current todo list from session messages by parsing TodoWrite tool calls.
    /// Returns the most recent state of todos (from the last TodoWrite call).
    /// </summary>
    IReadOnlyList<SessionTodoItem> ParseFromMessages(IReadOnlyList<ClaudeMessage> messages);
}

/// <summary>
/// Parses todo items from Claude session messages by extracting TodoWrite tool calls.
/// </summary>
public class TodoParser : ITodoParser
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    /// <inheritdoc />
    public IReadOnlyList<SessionTodoItem> ParseFromMessages(IReadOnlyList<ClaudeMessage> messages)
    {
        var allTodoWrites = messages
            .SelectMany(m => m.Content)
            .Where(c => c.Type == ClaudeContentType.ToolUse && c.ToolName == "TodoWrite")
            .ToList();

        if (allTodoWrites.Count == 0)
        {
            return [];
        }

        var lastTodoWrite = allTodoWrites[^1];

        if (string.IsNullOrEmpty(lastTodoWrite.ToolInput))
        {
            return [];
        }

        return ParseTodoJson(lastTodoWrite.ToolInput);
    }

    private static List<SessionTodoItem> ParseTodoJson(string json)
    {
        try
        {
            var wrapper = JsonSerializer.Deserialize<TodoWriteWrapper>(json, JsonOptions);
            if (wrapper?.Todos == null || wrapper.Todos.Count == 0)
            {
                return [];
            }

            return wrapper.Todos
                .Select(t => new SessionTodoItem
                {
                    Content = t.Content ?? string.Empty,
                    ActiveForm = t.ActiveForm ?? string.Empty,
                    Status = ParseStatus(t.Status)
                })
                .ToList();
        }
        catch (JsonException)
        {
            return [];
        }
    }

    private static TodoStatus ParseStatus(string? status)
    {
        return status?.ToLowerInvariant() switch
        {
            "pending" => TodoStatus.Pending,
            "in_progress" => TodoStatus.InProgress,
            "completed" => TodoStatus.Completed,
            _ => TodoStatus.Pending
        };
    }

    private class TodoWriteWrapper
    {
        public List<TodoItemJson>? Todos { get; set; }
    }

    private class TodoItemJson
    {
        public string? Content { get; set; }
        public string? ActiveForm { get; set; }
        public string? Status { get; set; }
    }
}

File: src/Homespun.Web/src/features/sessions/utils/todo-parser.ts
import type { AGUIMessage } from './agui-reducer'

export interface SessionTodoItem {
  content: string
  activeForm: string
  status: 'pending' | 'in_progress' | 'completed'
}

interface TodoWriteWrapper {
  todos?: Array<{
    content?: string
    activeForm?: string
    status?: string
  }>
}

/**
 * Extracts the current todo list from session messages by parsing TodoWrite tool calls.
 * Returns the most recent state of todos (from the last TodoWrite call).
 */
export function parseTodosFromMessages(messages: AGUIMessage[]): SessionTodoItem[] {
  const allTodoWrites = messages
    .flatMap((msg) => msg.content)
    .filter((block) => block.kind === 'toolUse' && block.toolName === 'TodoWrite')

  if (allTodoWrites.length === 0) {
    return []
  }

  const lastTodoWrite = allTodoWrites[allTodoWrites.length - 1]
  if (lastTodoWrite.kind !== 'toolUse' || !lastTodoWrite.input) {
    return []
  }

  return parseTodoJson(lastTodoWrite.input)
}

function parseTodoJson(json: string): SessionTodoItem[] {
  try {
    const wrapper: TodoWriteWrapper = JSON.parse(json)

    if (!wrapper.todos || wrapper.todos.length === 0) {
      return []
    }

    return wrapper.todos.map((todo) => ({
      content: todo.content || '',
      activeForm: todo.activeForm || '',
      status: parseStatus(todo.status),
    }))
  } catch {
    return []
  }
}

function parseStatus(status?: string): 'pending' | 'in_progress' | 'completed' {
  const normalized = status?.toLowerCase()

  switch (normalized) {
    case 'pending':
      return 'pending'
    case 'in_progress':
      return 'in_progress'
    case 'completed':
      return 'completed'
    default:
      return 'pending'
  }
}
